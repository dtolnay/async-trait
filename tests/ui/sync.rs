use async_trait::async_trait;

struct A;

#[async_trait]
pub trait Trait {
    async fn method(&self);
}

#[async_trait]
impl Trait for A {
    async fn method(&self) {}
}

fn main() {
    fn test<T>(a: T) where T: Trait + Send + Sync  {}
    test(A.method());
}
