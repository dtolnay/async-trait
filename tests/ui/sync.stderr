error[E0277]: the trait bound `Pin<Box<dyn Future<Output = ()> + Send>>: Trait` is not satisfied
  --> tests/ui/sync.rs:17:10
   |
17 |     test(A.method());
   |     ---- ^^^^^^^^^^ the trait `Trait` is not implemented for `Pin<Box<dyn Future<Output = ()> + Send>>`
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Trait` is implemented for `A`
note: required by a bound in `test`
  --> tests/ui/sync.rs:16:31
   |
16 |     fn test<T>(a: T) where T: Trait + Send + Sync  {}
   |                               ^^^^^ required by this bound in `test`

error[E0277]: `dyn Future<Output = ()> + Send` cannot be shared between threads safely
  --> tests/ui/sync.rs:17:10
   |
17 |     test(A.method());
   |     ---- ^^^^^^^^^^ `dyn Future<Output = ()> + Send` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Sync` is not implemented for `dyn Future<Output = ()> + Send`
   = note: required for `Unique<dyn Future<Output = ()> + Send>` to implement `Sync`
   = note: required because it appears within the type `Box<dyn Future<Output = ()> + Send>`
   = note: required because it appears within the type `Pin<Box<dyn Future<Output = ()> + Send>>`
note: required by a bound in `test`
  --> tests/ui/sync.rs:16:46
   |
16 |     fn test<T>(a: T) where T: Trait + Send + Sync  {}
   |                                              ^^^^ required by this bound in `test`
